Sources/NIOPosix/PosixSingletons+ConcurrencyTakeOver.swift

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Atomics
+import NIOCore
+
+#if compiler(>=5.9)
+private protocol SilenceWarning {
+    @available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+    func enqueue(_ job: UnownedJob)
+}
+@available(macOS 14, *)
+extension SelectableEventLoop: SilenceWarning {}
+#endif
+
+private let _haveWeTakenOverTheConcurrencyPool = ManagedAtomic(false)
+extension NIOSingletons {
+    /// Install ``MultiThreadedEventLoopGroup/singleton`` as Swift Concurrency's global executor.
+    ///
+    /// This allows to use Swift Concurrency and retain high-performance I/O alleviating the otherwise necessary thread switches between
+    /// Swift Concurrency's own global pool and a place (like an `EventLoop`) that allows to perform I/O
+    ///
+    /// This method uses an atomic compare and exchange to install the hook (and makes sure it's not already set). This unilateral atomic memory
+    /// operation doesn't guarantee anything because another piece of code could have done the same without using atomic operations. But we
+    /// do our best.
+    ///
+    /// - warning: You may only call this method from the main thread.
+    /// - warning: You may only call this method once.
+    @discardableResult
+    public static func unsafeTryInstallSingletonPosixEventLoopGroupAsConcurrencyGlobalExecutor() -> Bool {
+        #if compiler(>=5.9)
+        guard #available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *) else {
+            return false
+        }
+
+        typealias ConcurrencyEnqueueGlobalHook = @convention(thin) (
+            UnownedJob, @convention(thin) (UnownedJob) -> Void
+        ) -> Void
+
+        guard
+            _haveWeTakenOverTheConcurrencyPool.compareExchange(
+                expected: false,
+                desired: true,
+                ordering: .relaxed
```

Comment:
```suggestion
                ordering: .acquiringAndReleasing
```

This is probably safe without using an acquiring load here due to the `dlsym` call that follows it, but semantically this should be an acquiring load. Logically then it also has to be a releasing store because we're synchronising with ourselves!

Reply:
@Lukasa I think you're mixing up the two variables. This is just the guard, no other memory reads/writes are ordered with this guy. This is correct not because of the `dlsym` but because of the `guard`.

You need non-`.relaxed` if you care about the order of _other_ memory reads/writes but we don't

Reply:
Actually, we don't even need the `guard`. So yeah, any order of memory operations is completely fine here.